### PR TITLE
feat: expose the ability to use default encryption keys in fscloud submodule + DA

### PR DESCRIPTION
**Pull request by jor2**
### Description

[Expose provider visibility in the DA #320](https://github.com/terraform-ibm-modules/terraform-ibm-icd-elasticsearch/issues/320)
[Add workaround to ignore groups config when provisioning instance using backup CRN #324](https://github.com/terraform-ibm-modules/terraform-ibm-icd-elasticsearch/issues/324)
[Add support to use a different KMS key for backup encryption #323](https://github.com/terraform-ibm-modules/terraform-ibm-icd-elasticsearch/issues/323)
[Scope KMS policy to the exact KMS key #321](https://github.com/terraform-ibm-modules/terraform-ibm-icd-elasticsearch/issues/321)

### Release required?

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [x] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

##### Release notes content

Expose the ability to use default encryption keys in fscloud submodule + DA

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

### CODE REVIEW

**Comment by jor2** [comment]
based off this [pr](https://github.com/terraform-ibm-modules/terraform-ibm-icd-postgresql/pull/521)

**Comment by jor2** [comment]
Skipping upgrade tests, only seen this fail
````hcl
TestRunStandardUpgradeSolution 2024-11-11T20:49:28Z logger.go:66: │ Error: Value for undeclared variable
TestRunStandardUpgradeSolution 2024-11-11T20:49:28Z logger.go:66: │ 
TestRunStandardUpgradeSolution 2024-11-11T20:49:28Z logger.go:66: │ A variable named "provider_visibility" was assigned on the command line,
TestRunStandardUpgradeSolution 2024-11-11T20:49:28Z logger.go:66: │ but the root module does not declare a variable of that name. To use this
TestRunStandardUpgradeSolution 2024-11-11T20:49:28Z logger.go:66: │ value, add a "variable" block to the configuration.
TestRunStandardUpgradeSolution 2024-11-11T20:49:28Z logger.go:66: ╵
````

File: tests/pr_test.go [hunk 1]
```diff
@@ -193,6 +193,7 @@ func TestRunStandardSolutionSchematics(t *testing.T) {
 		{Name: "admin_pass_sm_secret_group", Value: options.Prefix, DataType: "string"},
 		{Name: "admin_pass_sm_secret_name", Value: options.Prefix, DataType: "string"},
 		{Name: "enable_kibana_dashboard", Value: true, DataType: "bool"},
+		{Name: "provider_visibility", Value: "public", DataType: "string"},
```
Comment (ocofaigh) [inline tests/pr_test.go:196]: no need to set to public in this test as schematics has access to private endpoints

File: modules/fscloud/variables.tf [hunk 2]
```diff
@@ -136,26 +136,33 @@ variable "auto_scaling" {
...
   type        = string
-  description = "The root key CRN of the Hyper Protect Crypto Services (HPCS) to use for disk encryption."
+  description = "The Hyper Protect Crypto Services (HPCS) or Key Protect root key CRN to use for disk encryption."
+  default     = null
 }
 
 variable "backup_encryption_key_crn" {
   type        = string
-  description = "The CRN of a Hyper Protect Crypto Services use for encrypting the disk that holds deployment backups. Only used if var.kms_encryption_enabled is set to true. There are limitation per region on the Hyper Protect Crypto Services and region for those services. See https://cloud.ibm.com/docs/cloud-databases?topic=cloud-databases-hpcs#use-hpcs-backups"
+  description = "The Hyper Protect Crypto Services (HPCS) or Key Protect root key CRN to use for encrypting the disk that holds deployment backups. There are region limitation for backup encryption. See https://cloud.ibm.com/docs/cloud-databases?topic=cloud-databases-hpcs#use-hpcs-backups (HPCS) and https://cloud.ibm.com/docs/cloud-databases?topic=cloud-databases-key-protect&interface=ui#key-byok (Key Protect)."
```
Comment (ocofaigh) [inline modules/fscloud/variables.tf:153]: region limitation -> region limitations (I fixed it in the postgres PR too :) )

**Review by ocofaigh — CHANGES_REQUESTED** [review]
some minor comments, and also please add the fix for https://github.com/terraform-ibm-modules/terraform-ibm-icd-elasticsearch/issues/324 as discussed

**Review by shemau — DISMISSED** [review]
LGTM. The restore/group changes look good to.

The schematics test failed.

**Comment by shemau** [comment]
```hcl
 2024/11/12 11:49:21 Terraform apply |   CreateSecretGroupWithContext failed: request_id: '77e1104f-3b52-a9fb-5499-4adc6343bb1c' path: 'manage/v2/groups' operation: 'create' a secret group with name 'els-sr-da-secret-group' already exists. [secrets-manager.04126E]
```

There seems to be a test collision if/when tests run in parallel the prefix is missing.

**Comment by ocofaigh** [comment]
@jor2 Still the same issue:
```
2024/11/12 16:18:21 Terraform apply |   CreateSecretGroupWithContext failed: request_id: 'bb136db9-014a-1488-9abf-ff03e560f000' path: 'manage/v2/groups' operation: 'create' a secret group with name 'els-sr-da-secret-group' already exists. [secrets-manager.04126E]
```

Somewhere a prefix value is not being passed and hence the secret name is clashing. Can you try find where and ensure prefix is being used

**Comment by ocofaigh** [comment]
Seems to be related to service credentials: `with module.secrets_manager_service_credentials[0].module.secret_groups["els-sr-da-secret-group"].ibm_sm_secret_group.secret_group,`

**Comment by ocofaigh** [comment]
FYI, there was also a bug in https://github.com/terraform-ibm-modules/terraform-ibm-icd-postgresql/pull/521 - I'm working on a fix so you will need that fix too

**Comment by jor2** [comment]
> Seems to be related to service credentials: `with module.secrets_manager_service_credentials[0].module.secret_groups["els-sr-da-secret-group"].ibm_sm_secret_group.secret_group,`

its because of this, any reason for it? https://github.com/terraform-ibm-modules/terraform-ibm-icd-elasticsearch/blob/main/tests/pr_test.go#L154

**Comment by ocofaigh** [comment]
FYI, there is currently an outage that is impacting tests

**Review by ocofaigh — APPROVED** [review]

**Review by shemau — APPROVED** [review]
LGTM

I focus on the restore/group logic and that is fine for now.

**Pull request merged**